### DISCUSSION TITLE
target zone vs ftp zone

### DISCUSSION THREAD

**Discussion by brokkolo** [cagnulein/qdomyos-zwift#2138] [Q&A] [answered]
since a few versions ago qz would use open training programs as reference for resistance level, thus change the resistance.
lately the beta version seems to be using the zone levels of the programs and changing the resistance accordingly.

my problem is that the target resistance is not aligned with the ftp zone, so a t. zone Z2.5 will try to keep me at Z1.9 and diminish the resistance if I try to reach that T.zone.
similarly t.zone Z3 keeps me at Z2.2
and T.zone of Z4.3 keeps me at Z2.8

this makes the programs unusable unless i want to do very easy spinning :(

I've tried resetting the app and also using default values.

toorx/iconsole bike with android app V.2.16.42

**Comment by cagnulein** [comment]
hi it's due to this
#1999
i will add a setting for this.
thanks for the feedback. it will be in the 2.16.45


Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno dom 18 feb 2024 alle 18:46 Brook ***@***.***> ha
scritto:

**Reply by brokkolo** [reply]
ok thank you 👍

**Comment by cagnulein** [comment]
done it will be in the 2.16.45

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno dom 18 feb 2024 alle ore 18:50 Brook ***@***.***>
ha scritto:

**Reply by brokkolo** [reply]
On 2.16.45
Changed toorx/iconsol options from toorx/iconsol bike to toorx srx 3500 but still cannot reach target zone as bike keeps lowering resistance to keep me in a lower zone.
I have a toorx srx 500
Should I do a spreadsheet with cadence/power/resistance data?
Or am I setting something wrong?

**Reply by cagnulein** [reply]
@brokkolo you don't have to change any settings (restore the previous ones). i just reverted the modification so it will be the same as before

**Reply by brokkolo** [reply]
Exactly, now I'm back to a situation where the resistance doesn't change when I increase or decrease cadence.

If I use the srx 3500 setting then the resistance changes when I change cadence, but the change coefficient is not right for the srx 500, keeping target zone is always higher than actual zone.

**Comment by cagnulein** [comment]
hah ok so i thought you were happy about the original situation, so it
isn't right?
so yea i need a power/cadence/resistance table from your bike
let me know!

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno dom 25 feb 2024 alle 13:09 Brook ***@***.***> ha
scritto:

**Reply by brokkolo** [reply]
Perfetto!!
Will try to prepare a chart later today.
Tnx

**Reply by brokkolo** [reply]
not as neat as the other chart :)

[toorx srx 500 cadence.resistance.power.xlsx](https://github.com/cagnulein/qdomyos-zwift/files/14397128/toorx.srx.500.cadence.resistance.power.xlsx)

**Comment by cagnulein** [comment]
perfect i will give it a go!

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno dom 25 feb 2024 alle 18:09 Brook ***@***.***> ha
scritto:

**Comment by brokkolo** [comment]
if you need more data this is the stripped down fit file
[output_2024.02.25 20.20.43_c42bb155.xlsx](https://github.com/cagnulein/qdomyos-zwift/files/14397899/output_2024.02.25.20.20.43_c42bb155.xlsx)

**Comment by cagnulein** [comment]
Awesome, thanks!


Il giorno dom 25 feb 2024 alle 22:29 Brook ***@***.***> ha
scritto:

**Comment by cagnulein** [comment]
done @brokkolo 
https://github.com/cagnulein/qdomyos-zwift/actions/runs/8046012010
Google is very slow in these days, reviewing the app on the store, even the beta ones. For this reason, in order to check right now if the patch that I did is right, you can use the android build in the link (check at the bottom of the page in about 1 hour). In order to test it you need first to remove the current QZ version on your phone, install the one that you downloaded.
This version is a 15 minutes trial (you can restart it as many times as you want) and I can unlock it if you want, send me an email to roberto.viola83@gmail.com mentioning this ticket and that you are on android.
Let me know if you have any questions

**Reply by brokkolo** [reply]
Is this done in 2.16.46?
I tried with the trial build and the "enerfit spx 9500/toorx srx 500" setting doesn't change resistance when increasing cadence like srx 3500 does :(

**Reply by cagnulein** [reply]
yes you have to disable the toorx_srx_3500 setting and enable the erg mode of course. Could you please share a debug log in this scenario?

**Reply by cagnulein** [reply]
hah sorry i forgot a line!

**Reply by cagnulein** [reply]
new build https://github.com/cagnulein/qdomyos-zwift/actions/runs/8052225015

**Reply by brokkolo** [reply]
👍 this build works! Thank you

The zones still don't match 100% but much better than srx 3500.

Is there anything I can provide to make them match better?

**Comment by cagnulein** [comment]
Can you give me an example of the missing precision? I can try to tell to
chat gpt to improve it

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno lun 26 feb 2024 alle 18:44 Brook ***@***.***> ha
scritto:

**Reply by brokkolo** [reply]
excellent question that got me thinking a bit... if i understand what you are doing is looking for a function, but in my case this function seems to be more reliable with cadences around 60... so when i go up or down is starts getting out of sync.
I'm not a cyclist but a runner and have no clue on what good cadence is, and being a runner i like fast cadences as it seems to better mimic running pace (something around 80)
So I guess i would like the formula to be more aligned with cadence between 60 and 100, with 80 as a sweet spot.
so to build the formula we could avoid trying to match unneeded resistances/cadences, so using only something like:

 - for lower resistances (between 5 and 10) we could consider the data that has cadence between 70 and 110
 - for medium resistance (between 10 and 20) consider data with cadence between 60 and 100
 - for high resistance (20 up) consider data with cadence between 50 and 90
 
 this might make the formula more useful...
 
 Spero tanto di essermi spiegato :)
 
 here is a bit more data (perhaps duplicate data needs to be removed as the formula might also take into consideration how many times certain data appears?)
 
 
[toorx new.xlsx](https://github.com/cagnulein/qdomyos-zwift/files/14422210/toorx.new.xlsx)

**Reply by cagnulein** [reply]
i added the file to chatgpt https://chat.openai.com/share/6a13b8ab-b8ac-4bdc-92a0-817709a7ed53
try to check if it's better asking him to give you the output wattage from a certain cadence and resistance

**Reply by brokkolo** [reply]
Sorry but I'm unable to continue your chatgpt conversation, it sends me to a new chat.

Quick question: I think I'm using the default option of letting QZ do the power calculation and not reading the power from the bike. Should I turn that option off?

**Reply by cagnulein** [reply]
@brokkolo the default settings are right. The model provided from chatgpt is always underestimate the power for the 80 bpm cadence. Maybe we can try to clean the excel file and keep only the cadence between 70-90. I will try to do this in the next days

**Reply by brokkolo** [reply]
i've increased the sample hoping for better fit, cleaned up the data removing anything below 125w to eliminate low cadences, and removed duplicates.

fingers crossed and thanx as always :)
[toorx new - filtered starting 125w.xlsx](https://github.com/cagnulein/qdomyos-zwift/files/14430132/toorx.new.-.filtered.starting.125w.xlsx)

**Reply by cagnulein** [reply]
@brokkolo can you check if this it's any better? 
[combined_polynomial_function.zip](https://github.com/cagnulein/qdomyos-zwift/files/14437606/combined_polynomial_function.zip)

**Reply by brokkolo** [reply]
hi again, so, i'm not good with python thus i tried a few solutions on my own:
downloaded all fit files from qz and put them in a table, filtered inconsistent data and duplicates.
then i tried your road of finding a formula for each resistance separately, it seems better than before.
used this https://stats.blue/Stats_Suite/polynomial_regression_calculator.html setting it to
y=β0+β1x+β2x^2+β3x^3 with 10 decimals
here is the collected data and some of the regressions divided by resistance.
[all qz table with date filtered.xlsx](https://github.com/cagnulein/qdomyos-zwift/files/14465050/all.qz.table.with.date.filtered.xlsx)


unfortunately as the resistance gets harder the power curve increases a bend both on 60 and 80??
this is with a resistance of 20
![scatter_plot](https://github.com/cagnulein/qdomyos-zwift/assets/73337253/7cbc8302-b92b-4786-baa8-931cf11af2d5)

so here is another question: when i'm cycling on the qz interface i see both the target power and the actual power. why does the app keep trying to not let me reach the target power? doesn't it read the actual power (the one seen on the interface) and adjust resistance accordingly? why do we need a new formula if the current power is already shown, can't we use that?

What i'm trying to get to is that the current power shown on qz is ok for me, i would like the auto resistance to take that power data into consideration when adjusting.

**Reply by cagnulein** [reply]
qz know the target and the actual power but how can know the conversion to the resistance :) that's why we need a power table. it's how all the fitness devices work :) 

so are you satisfied with your table ? could i implement it?

**Comment by brokkolo** [comment]
> qz know the target and the actual power but how can know the conversion to the resistance :) that's why we need a power table. it's how all the fitness devices work :)

ah, ok i thought you could just program qz to increase resistance by one if target power was 10 points higher or reduce if lower... and adjust every few seconds... without any need to know the conversion. this would be more simple and wouldn't need info about cadence either.

> so are you satisfied with your table ? could i implement it?

the table needs finishing, some of the higher resistances are missing, will have to do some tough sessions in the next days.
will update asap :)

**Reply by cagnulein** [reply]
> ah, ok i thought you could just program qz to increase resistance by one if target power was 10 points higher or reduce if lower... and adjust every few seconds... without any need to know the conversion. this would be more simple and wouldn't need info about cadence either.

This will lead in strange jumps for sure :) 

If you want, instead of building this table, I'm working on a creating a dynamic power table. in this way we would need to create a power table at all. it will get the table in real time. what do you think? i will finalize it in the next days

**Reply by brokkolo** [reply]
Dynamic power table sounds very interesting, happy to test it when it's ready.
I will try to complete the missing high resistance data anyhow, just in case it can be useful 😀

**Comment by cagnulein** [comment]
sure we can do both things!

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno sab 2 mar 2024 alle ore 07:57 Brook ***@***.***> ha
scritto:

**Reply by cagnulein** [reply]
@brokkolo done! https://github.com/cagnulein/qdomyos-zwift/actions/runs/8121336051

what I did is adding a way to fill a dynamic table. So first of all enable the debug log with this build. Then spin a little to every caadence/resistance in order to fill it. Then try do the erg. I didn't test my self so we need some version to stabilize it. I did a lot of code for this so be careful :D 
let me know!

**Reply by brokkolo** [reply]
Yes!!! This works very well!
You rock my friend 👍

**Reply by cagnulein** [reply]
Really? @brokkolo  It's work? That's amazing! Could you please send me a debug log anyway from a session?

**Reply by brokkolo** [reply]
[debug-Sun_Mar_3_11_16_37_2024.log](https://github.com/cagnulein/qdomyos-zwift/files/14472604/debug-Sun_Mar_3_11_16_37_2024.log)
[debug-Sun_Mar_3_11_15_47_2024.log](https://github.com/cagnulein/qdomyos-zwift/files/14472605/debug-Sun_Mar_3_11_15_47_2024.log)
Not sure which of these is good

I can do a new one if you need

**Reply by cagnulein** [reply]
Actually they are very short sessions @brokkolo 

Can you send me a bigger one when you will do it?

Thanks

**Reply by brokkolo** [reply]
Here is a bit of messing around
[debug-Sun_Mar_3_12_19_49_2024.log](https://github.com/cagnulein/qdomyos-zwift/files/14472714/debug-Sun_Mar_3_12_19_49_2024.log)

Perhaps still a bit  jumpy but always trying to get me to the right power. Good
I guess the database needs to have much more data recorded
A few questions: 
Does it only use power recorded in database or also will it try data not in database?
Does it only fill database when not in erg mode? 
Also, does the database get erased at restart?
Finally: I'm using both toorx/iconsol and toorx srx500, should I only use one of them?

**Reply by cagnulein** [reply]
it also interpolate data @brokkolo

**Reply by cagnulein** [reply]
the db is saved. anyway i'm seeing some strange things in the log, so a fix will be required

leave the settings as they are

**Reply by cagnulein** [reply]
@brokkolo i did a modification to fixing some nan and inf situation https://github.com/cagnulein/qdomyos-zwift/actions/runs/8131624448

it should be better. please provide a new debug log when you will try. thanks!

**Reply by brokkolo** [reply]
Much better!
Got 2 logs: 
1st one is with erg off creating dynamic database (cannot install new app version onto of old, need to uninstall old one first, losing all settings and database)
[debug-Mon_Mar_4_08_38_02_2024.log](https://github.com/cagnulein/qdomyos-zwift/files/14480076/debug-Mon_Mar_4_08_38_02_2024.log)

2nd one with erg on doing the "short mix ..." training program.
[debug-Mon_Mar_4_08_53_42_2024.log](https://github.com/cagnulein/qdomyos-zwift/files/14480087/debug-Mon_Mar_4_08_53_42_2024.log)

Now it's not as jumpy!
tried varying cadences and also going slow to have qz use higher resistances that I hadn't recorded in DB, and qz did always a good job in keeping actual power in line with target power 👏

**Reply by cagnulein** [reply]
> 1st one is with erg off creating dynamic database (cannot install new app version onto of old, need to uninstall old one first, losing all settings and database)

yes until i will not merge this, a reinstall is always required. actually you can create the table while you're doing the erg mode. it will build it dinamically!

you can check the log about the table filtering for "newPointAdded". that's so cool!

> Now it's not as jumpy!
tried varying cadences and also going slow to have qz use higher resistances that I hadn't recorded in DB, and qz did always a good job in keeping actual power in line with target power 👏

perfect, I just add another modification to improve it. I will send you the link shortly

**Reply by cagnulein** [reply]
@brokkolo new build https://github.com/cagnulein/qdomyos-zwift/actions/runs/8139136294
anyway as you already did give it a spin to all the useful resistance to fill the table

we are near to the right one if it's also ok this one

**Comment by brokkolo** [comment]
> you can create the table while you're doing the erg mode. it will build it dinamically!
> 
Fantastic
> you can check the log about the table filtering for "newPointAdded". that's so cool!
> 
👍

How about an experimantal option to export/import table?

**Reply by cagnulein** [reply]
> How about an experimantal option to export/import table?

you will be able to do this from the next version saving the setting and importing them (or the profile)

**Comment by cagnulein** [comment]
with the help of @drmason789 I also added your table into the test case so I should see the estimate wattage directly without doing a real test :)

**Comment by cagnulein** [comment]
@brokkolo wait to test the newer version. i'm fixing some edge cases

**Comment by cagnulein** [comment]
@brokkolo use this build https://github.com/cagnulein/qdomyos-zwift/actions/runs/8142359641

**Comment by brokkolo** [comment]
2 logs, first skipping through some resistances and cadences, 2nd doing a ftp training (had to zip it as too big for upload)

All ok as before, gets a little jumpy with higher powers as I guess variability increases so for example at 340pw a change of 1 in resistance could mean +/- 40power.

[debug-Mon_Mar_4_17_03_52_2024.log](https://github.com/cagnulein/qdomyos-zwift/files/14485068/debug-Mon_Mar_4_17_03_52_2024.log)
[debug-Mon_Mar_4_17_15_12_2024.zip](https://github.com/cagnulein/qdomyos-zwift/files/14485087/debug-Mon_Mar_4_17_15_12_2024.zip)


So all good for me!

One question: is it normal that at same power, a high cadence (100) feels much "easier" than a low one (50)???

**Reply by cagnulein** [reply]
perfect! i will check the log tomorrow. I saw in your past logs that sometimes some values of wattage were the same on different resistance value (for example 150W for cad70 at res 12 and 13). I guess because the bike took a little to update the wattage. I'm wondering if i have filter out this...what do you think? I have an idea about how I can do this.

> One question: is it normal that at same power, a high cadence (100) feels much "easier" than a low one (50)???

yes if it's the same resistance level

**Comment by brokkolo** [comment]
>I saw in your past logs that sometimes some values of wattage were the same on different resistance value (for example 150W for cad70 at res 12 and 13). I guess because the bike took a little to update the wattage. I'm wondering if i have filter out this...what do you think? 
> 
What I did with my table was to remove the od data by keeping the most frequent ones, but this means that you need to keep ALL the data so you can see which is more frequent.
Another way could be deleting a data if, at same resistance, power is not higher than previous cadence or lower than next cadence... 
 
> > One question: is it normal that at same power, a high cadence (100) feels much "easier" than a low one (50)???
> 
> yes if it's the same resistance level

No, I actually mean something like:
 340w at resistance 10 with a cadence of 100 is soooo much easier than 340w at resistance 25 at cadence 50... (numbers are just to give an idea). Is this normal ??

**Reply by cagnulein** [reply]
it depends from your muscle and bike :) i guess it's the same even without qz, right?

**Reply by brokkolo** [reply]
You're right, it's a [muscle thing](https://www.trainingpeaks.com/blog/low-cadence-intervals-cycling/)

**Comment by cagnulein** [comment]
ok, note for me. Check why there was a resisstance change here

```
Mon Mar 4 18:00:19 2024 1709571619535 Debug: trainprogram.cpp void trainprogram::scheduler() "trainprogram change power 338"
Mon Mar 4 18:00:19 2024 1709571619663 Debug: devices/bluetooth.cpp void bluetooth::debug(const QString &) "Current cadence: 101"
Mon Mar 4 18:00:19 2024 1709571619666 Debug: devices/bluetooth.cpp void bluetooth::debug(const QString &) "Current watt: 344.2"
Mon Mar 4 18:00:20 2024 1709571620163 Debug: devices/bluetooth.cpp void bluetooth::debug(const QString &) "Current cadence: 100"
Mon Mar 4 18:00:20 2024 1709571620167 Debug: devices/bluetooth.cpp void bluetooth::debug(const QString &) "Current watt: 302"
Mon Mar 4 18:00:20 2024 1709571620537 Debug: trainprogram.cpp void trainprogram::scheduler() "trainprogram change power 338"
Mon Mar 4 18:00:20 2024 1709571620681 Debug: devices/bluetooth.cpp void bluetooth::debug(const QString &) "Current cadence: 100"
Mon Mar 4 18:00:20 2024 1709571620685 Debug: devices/bluetooth.cpp void bluetooth::debug(const QString &) "Current watt: 302"
Mon Mar 4 18:00:20 2024 1709571620693 Debug: devices/bluetooth.cpp void bluetooth::debug(const QString &) "writing resistance 11"
Mon Mar 4 18:00:21 2024 1709571621204 Debug: devices/bluetooth.cpp void bluetooth::debug(const QString &) "Current cadence: 97"
Mon Mar 4 18:00:21 2024 1709571621208 Debug: devices/bluetooth.cpp void bluetooth::debug(const QString &) "Current watt: 287.3"
```

**Comment by cagnulein** [comment]
@brokkolo you can also use big numbers in the zwift erg filter up and down. for example I could suggest a 20 and 20 for you. Let me know

**Reply by brokkolo** [reply]
Yup, will try increasing to 15. But im happy as it is. The issue is more on the extremes of the power spectrum

**Reply by cagnulein** [reply]
yes the 15 or 20 should help you in that power spectrum.
I just add a patch for filtering the table when the resistance change (i added a 1s timeout) in order to don't fill the table with incongruent datas

https://github.com/cagnulein/qdomyos-zwift/actions/runs/8152418643

**Reply by cagnulein** [reply]
@brokkolo did you check also the last build?

**Reply by brokkolo** [reply]
been running lately. will give it a spin tomorrow ;)

**Reply by cagnulein** [reply]
@brokkolo @drmason789 i'm also wondering if it could have sense to don't add samples under a minimum cadence value, for example 40?

**Reply by brokkolo** [reply, marked as answer]
Love it 😀 
The app is working well for me, didn't even do a spinning to get the database filled up, just started a training and it was quite accurate (had +-15 instead of 10).
Log just in case you need it
[debug-Thu_Mar_7_10_33_35_2024.zip](https://github.com/cagnulein/qdomyos-zwift/files/14523907/debug-Thu_Mar_7_10_33_35_2024.zip)

For me you can filter out under 40... unless someone wants to do very high resistance to mimic going uphill on a mtb?!?

Do the test builds not send the ride via email? Wanted to post the graph of the ride to show you how accurate it was compared to the training but qz didn't send me the ride.

**Reply by cagnulein** [reply]
> Love it 😀
> The app is working well for me, didn't even do a spinning to get the database filled up, just started a training and it was quite accurate (had +-15 instead of 10).

amazing i'm checking the log and also the new way to avoid to fill the table with spurious value due to the resistance changing it's working pretty well!

> For me you can filter out under 40... unless someone wants to do very high resistance to mimic going uphill on a mtb?!?

ok let leave this as it is for now, I could always change this in the future

> Do the test builds not send the ride via email? Wanted to post the graph of the ride to show you how accurate it was compared to the training but qz didn't send me the ride.

yes it doesn't. I'm merging this now, removing some debug and then I will push it to the play store version in beta!

Thanks for the feedbacks!

**Reply by cagnulein** [reply]
just a question @brokkolo are you still loading the settings when you're loading the new apk, right?

**Reply by brokkolo** [reply]
I copied the .qzs profile from the previous test build to not have to go through all the settings again

**Comment by cagnulein** [comment]
perfect it explains some dirty lines! thanks for the confirmation.

code merged!

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno gio 7 mar 2024 alle 15:58 Brook ***@***.***> ha
scritto:

**Reply by brokkolo** [reply]
My setup is:
- Toorx srx 500
- Garmin hrm belt
Qz reads from both.
Qz sends cadence, power and speed via ant to garmin watch wich also is connected to hrm belt.
Something I noticed is that before qz would slightly underestimate calories compared to garmin watch. Now it has reversed and qz is overestimating calories by about 1/4
Only thing that has changed in settings is that before I had "toorx/iconsole bike" setting checked, now I have "enerfit ... /toorx srx 500" setting.

**Reply by cagnulein** [reply]
@brokkolo qz is calculating calories from wattage so, if the calibration of the bike didn't change, the power value and so the calories should be the same. also age and gender affects calories calculation on qz
